### PR TITLE
Improve logging in error-handler

### PR DESCRIPTION
**Pull request by scepticulous**
We received multiple error reports that did not contain a sentry
reference and if it did, the information we were able was not sufficient
to debug this issue.
This change addresses the issue by logging more details in case of an internal error that is reported to the user via an ephemeral message.

### CODE REVIEW

File: lib/error-handler.js [hunk 1]
```diff
@@ -36,20 +37,40 @@ module.exports = {
...
 
     // eslint-disable-next-line no-unused-vars
-    app.use((err, req, res, next) => {
-      req.log.error(err);
+    app.use((err, req, res, _next) => {
+      const customUUID = uuidv4();
+      const isInSentry = res.sentry !== undefined;
 
+      if (!res.sentry) {
+        // create a reference uuid if sentry did not catch the error
```
[the remark this answers is not in the document]
Comment (scepticulous) [reply lib/error-handler.js]: That is what we are trying to investigate.
We have incoming support tickets with 5xx errors that have undefined values for `.sentry` . This PR is part of an investigation to understand the cause and to ensure we stop having incoming support tickets without information.

File: lib/middleware/route-command.js [hunk 2]
```diff
@@ -38,7 +38,7 @@ module.exports = function route(req, res, next) {
     // Slack has a 3000ms timeout, so if we don't respond within that window, ack
     // the command and use the delayed response URL.
     timeout = setTimeout(() => {
-      req.log('Command taking too long. Switching to delayed response.');
+      req.log.warn('Command taking too long. Switching to delayed response.');
```
[the remark this answers is not in the document]
Comment (scepticulous) [reply lib/middleware/route-command.js]: It should end up in `loggly` with syslog level of 'warn', which is also what I am aiming for in general.

File: lib/error-handler.js [hunk 3]
```diff
@@ -36,20 +56,35 @@ module.exports = {
...
 
     // eslint-disable-next-line no-unused-vars
-    app.use((err, req, res, next) => {
-      req.log.error(err);
+    app.use((err, req, res, _next) => {
+      const customUUID = uuidv4();
 
+      if (!res.sentry) {
+        // create a reference uuid if sentry did not catch the error
+        res.sentry = customUUID;
```
[the remark this answers is not in the document]
Comment (scepticulous) [reply lib/error-handler.js]: I was surprised as well, However, this PR is intended to be backward compatible to ship it with minimal risk.
The confusion will be addressed when I make the step of replacing Raven with the new sentry sdk.
That's why I did not alter the confusing naming.

File: lib/error-handler.js [hunk 4]
```diff
@@ -36,20 +56,35 @@ module.exports = {
...
       if (command) {
         return command.respond(message);
       } else if (res.headersSent) {
-        req.log.warn({ err }, 'Headers were already sent, so swallowing error');
+        req.log.warn({ err, ref: res.sentry }, 'Headers were already sent. Not responding to user.');
       } else if (isRequestFromSlack(req)) {
         res.json(message);
       } else {
-        res.status(500).send(`An unexpected error occured: ${res.sentry}`);
+        res.status(500).send(`An unexpected error occured. ref: ${res.sentry}`);
```
[the remark this answers is not in the document]
Comment (scepticulous) [reply lib/error-handler.js]: We are already logging at line 76 and include a `ref` there.

**Comment by scepticulous** [comment]
> Do you have references to the errors mentioned in the description? I don't fully understand the condition this aims to address.

We have 5xx reports, with `res.sentry` being undefined, which results in support issues that have a text that includes: `Error code: undefined`.
Generally, I don't care why sentry/Raven is unable to provide an id for these errors (Too many possibilities). Instead, I want to make sure, that our log provider can draw a complete picture for all errors without relying on the (outdated/unmaintained) sentry middleware we are using.

File: lib/error-handler.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (scepticulous) [reply lib/error-handler.js]: 😄 good catch. I guess that refactoring broke everything 😆

File: lib/error-handler.js [hunk 5]
```diff
@@ -12,10 +26,16 @@ module.exports = {
...
+    const timeoutValue = '10s';
+    app.use(timeout(timeoutValue));
     app.use((req, res, next) => {
       req.on('timeout', () => {
-        req.log.warn('Request timeout');
+        const errorInfo = {
+          ...defaultErrorInfo,
+          isTimeout: true,
+        };
+        req.log.warn(errorInfo, `Request timed out after ${timeoutValue}.`);
```
Comment (jnraine) [inline lib/error-handler.js]: Is it worth responding with a `502` as well to close the connection or do you want the request to continue out?

File: lib/error-handler.js [hunk 6]
```diff
@@ -12,10 +26,16 @@ module.exports = {
     }).install();
 
     app.use(Raven.requestHandler());
-    app.use(timeout('10s'));
+
+    const timeoutValue = '10s';
+    app.use(timeout(timeoutValue));
```
Comment (jnraine) [inline lib/error-handler.js]: The [docs](https://www.npmjs.com/package/connect-timeout) for `connect-timeout` add this caveat:
 
> While the library will emit a 'timeout' event when requests exceed the given timeout, node will continue processing the slow request until it terminates.

Is it okay that this won't stop processing the request?

File: lib/error-handler.js [hunk 7]
```diff
@@ -36,20 +56,30 @@ module.exports = {
...
       const { command } = res.locals;
-      const message = new Exception(res.sentry, command).toJSON();
+      const ref = (res.sentry || customRef);
+
+      const errorInfo = {
+        ...defaultErrorInfo(req, res),
+        ref,
+        customRef,
+        error,
+        isTimeout: false,
```
Comment (jnraine) [inline lib/error-handler.js]: 💯

File: lib/error-handler.js (hunk 6 above)
Comment (scepticulous) [reply lib/error-handler.js]: For now, I don't want to alter the behavior. 
The `next` middleware is actually adding an on-timeout handler.
For now I only want to get more insights, but eventually, I will start aborting the process.

File: lib/error-handler.js (hunk 5 above)
Comment (scepticulous) [reply lib/error-handler.js]: The integration is a little complicated in that regard. For certain things, we immediately respond to the client and just continue processing the request. In other cases, we are working synchronously and for those cases that is an option.
However, error handling is something I'll improve later and this PR should pave the path by allowing better insights.

**Pull request merged**